### PR TITLE
[orch] change Consumer class to support multiple values for the same key

### PR DESCRIPTION
**Pull request by zhenggen-xu**
**Description**

The Consumer class is used by the orch objects to deal with the redis consumer tables' popped items. It has m_toSync map to save the tasks.  During the operations (more items than the orch objects can handle), the tasks in the map is merged for optimization. However, since it is a map, we would only have one value for each key.  This potentially eliminate the necessary actions from Redis, e,g, we have a DEL action and SET action coming out from Redis for some key, we would overwrite the DEL action today in the code and thus not able to delete some objects as we intended. 

The PR changed the m_toSync  from map to multi-map to get multiple values for one key.  In this design, we keep maximun two values per key,  DEL or SET or DEL+SET.  We need strictly keep the order of DEL and SET.  It is possible to use map of vectors to fulfill this,  we chose multi-map because:
 1,  It will have less/no changes to different orch classes to iterate the m_toSync 
 2, The order can be guaranteed. The order of the key-value pairs whose keys compare equivalent is the order of insertion and does not change. (since C++11). See
  https://en.cppreference.com/w/cpp/container/multimap

The PR also refactors the consumer class so vlanmgr.cpp and routeorch.cpp will leverage the Consumer functions instead of operating on the members.  It also refactors the UT code (aclorch_ut.cpp) so it removes the redundant code and uses the same code.

Google UT tests were added for Consumer Class especially for different cases for addToSync() function.

**What I did**
Change the m_toSync in Consumer class to multimap so it could support both DEL and SET
Reload Consumer addToSync() and refactor vlanmgr/route-orch and ut code to use it
Add google ut for consumer class
    
**Why I did it**
See description.

**How I verified it**
Unit tests:
Running main() from gtest_main.cc
```
[==========] Running 19 tests from 5 test cases.
[----------] Global test environment set-up.
[----------] 1 test from AclTest
[ RUN      ] AclTest.Create_L3_Acl_Table
[       OK ] AclTest.Create_L3_Acl_Table (1 ms)
[----------] 1 test from AclTest (1 ms total)

[----------] 3 tests from AclOrchTest
[ RUN      ] AclOrchTest.ACL_Creation_and_Destorying
[       OK ] AclOrchTest.ACL_Creation_and_Destorying (1000 ms)
[ RUN      ] AclOrchTest.L3Acl_Matches_Actions
[       OK ] AclOrchTest.L3Acl_Matches_Actions (1001 ms)
[ RUN      ] AclOrchTest.L3V6Acl_Matches_Actions
[       OK ] AclOrchTest.L3V6Acl_Matches_Actions (1000 ms)
[----------] 3 tests from AclOrchTest (3003 ms total)

[----------] 2 tests from PortsOrchTest
[ RUN      ] PortsOrchTest.PortReadinessColdBoot
[       OK ] PortsOrchTest.PortReadinessColdBoot (21 ms)
[ RUN      ] PortsOrchTest.PortReadinessWarmBoot
[       OK ] PortsOrchTest.PortReadinessWarmBoot (13 ms)
[----------] 2 tests from PortsOrchTest (34 ms total)

[----------] 4 tests from SaiSpy
[ RUN      ] SaiSpy.CURD
[       OK ] SaiSpy.CURD (0 ms)
[ RUN      ] SaiSpy.Same_Function_Signature_In_Same_API_Table
[       OK ] SaiSpy.Same_Function_Signature_In_Same_API_Table (0 ms)
[ RUN      ] SaiSpy.Same_Function_Signature_In_Different_API_Table
[       OK ] SaiSpy.Same_Function_Signature_In_Different_API_Table (0 ms)
[ RUN      ] SaiSpy.create_switch_and_acl_table
[       OK ] SaiSpy.create_switch_and_acl_table (0 ms)
[----------] 4 tests from SaiSpy (0 ms total)

[----------] 9 tests from ConsumerTest
[ RUN      ] ConsumerTest.ConsumerAddToSync_Set
[       OK ] ConsumerTest.ConsumerAddToSync_Set (1 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Del
[       OK ] ConsumerTest.ConsumerAddToSync_Del (0 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Set_Del
[       OK ] ConsumerTest.ConsumerAddToSync_Set_Del (0 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Del_Set
[       OK ] ConsumerTest.ConsumerAddToSync_Del_Set (130 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Set_Del_Set_Multi
[       OK ] ConsumerTest.ConsumerAddToSync_Set_Del_Set_Multi (204 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Set_Del_Set_Multi_In_Q
[       OK ] ConsumerTest.ConsumerAddToSync_Set_Del_Set_Multi_In_Q (4 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Del_Set_Setnew
[       OK ] ConsumerTest.ConsumerAddToSync_Del_Set_Setnew (0 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Del_Set_Setnew1
[       OK ] ConsumerTest.ConsumerAddToSync_Del_Set_Setnew1 (0 ms)
[ RUN      ] ConsumerTest.ConsumerAddToSync_Ind_Set_Del
[       OK ] ConsumerTest.ConsumerAddToSync_Ind_Set_Del (0 ms)
[----------] 9 tests from ConsumerTest (340 ms total)

[----------] Global test environment tear-down
[==========] 19 tests from 5 test cases ran. (4344 ms total)
[  PASSED  ] 19 tests.
```

**Details if related**

### CODE REVIEW

File: cfgmgr/vlanmgr.cpp [hunk 1]
```diff
@@ -447,8 +447,9 @@ void VlanMgr::processUntaggedVlanMembers(string vlan, const string &members)
             vector<FieldValueTuple> fvVector;
             FieldValueTuple t("tagging_mode", "untagged");
             fvVector.push_back(t);
-            consumer.m_toSync[member_key] = make_tuple(member_key, SET_COMMAND, fvVector);
-            SWSS_LOG_DEBUG("%s", (dumpTuple(consumer, consumer.m_toSync[member_key])).c_str());
+            KeyOpFieldsValuesTuple tuple =  make_tuple(member_key, SET_COMMAND, fvVector);
```
Comment (qiluo-msft) [inline cfgmgr/vlanmgr.cpp]: >   [](start = 42, length = 2)

Use one blank

File: orchagent/orch.cpp [hunk 2]
```diff
@@ -67,36 +67,66 @@ vector<Selectable *> Orch::getSelectables()
...
+    }
+    else
+    {
+        /*
+        * Now we are trying to add the key-value with SET.
+        * We maintain maximun two values per key.
+        * In case there is one key-value, it should be DEL or SET
+        * In case there are two key-value pairs, it should be DEL then SET
+        * The code logic is following:
+        * We iterating the values with the key, we skip the value with DEL and then
```
Comment (qiluo-msft) [inline orchagent/orch.cpp]: >iterating [](start = 13, length = 9)

iterate

File: orchagent/orch.cpp [hunk 3]
```diff
@@ -67,36 +67,66 @@ vector<Selectable *> Orch::getSelectables()
...
         {
-            Orch::recordTuple(*this, entry);
+            auto old_op = kfvOp(iter->second);
+            if (old_op == SET_COMMAND)
+                break;
         }
-
-        /* If a new task comes or if a DEL task comes, we directly put it into getConsumerTable().m_toSync map */
-        if (m_toSync.find(key) == m_toSync.end() || op == DEL_COMMAND)
+        if (iter == m_toSync.end() || iter->first != key)
```
Comment (qiluo-msft) [inline orchagent/orch.cpp]: >iter == m_toSync.end() || iter->first != key [](start = 12, length = 44)

iter == ret.second

File: orchagent/orch.cpp [hunk 4]
```diff
@@ -118,9 +148,22 @@ size_t Consumer::addToSync(std::deque<KeyOpFieldsValuesTuple> &entries)
                 }
                 existing_values.push_back(FieldValueTuple(field, value));
             }
-            m_toSync[key] = KeyOpFieldsValuesTuple(key, op, existing_values);
+            m_toSync.erase(iter);
+            m_toSync.emplace(key, KeyOpFieldsValuesTuple(key, op, existing_values));
```
Comment (qiluo-msft) [inline orchagent/orch.cpp]: Is it possible to directly write
```
iter->second = KeyOpFieldsValuesTuple(key, op, existing_values);
```

File: orchagent/orch.h [hunk 5]
```diff
@@ -143,9 +147,11 @@ class Consumer : public Executor {
     // TODO: hide?
     SyncMap m_toSync;
 
+    void addToSync(const swss::KeyOpFieldsValuesTuple &entry);
+
 protected:
     // Returns: the number of entries added to m_toSync
-    size_t addToSync(std::deque<swss::KeyOpFieldsValuesTuple> &entries);
+    size_t addToSync(const std::deque<swss::KeyOpFieldsValuesTuple> &entries);
```
Comment (qiluo-msft) [inline orchagent/orch.h]: >addToSync [](start = 11, length = 9)

It make more sense to be public.

File: orchagent/routeorch.cpp [hunk 6]
```diff
@@ -405,8 +405,8 @@ void RouteOrch::doTask(Consumer& consumer)
                     {
                         vector<FieldValueTuple> v;
                         key = vrf + i.first.to_string();
-                        auto x = KeyOpFieldsValuesTuple(key, DEL_COMMAND, v);
-                        consumer.m_toSync[key] = x;
+                        KeyOpFieldsValuesTuple x = KeyOpFieldsValuesTuple(key, DEL_COMMAND, v);
```
Comment (qiluo-msft) [inline orchagent/routeorch.cpp]: `auto` is better

**Comment by lguohan** [comment]
i am ok with this pr.

File: cfgmgr/vlanmgr.cpp (hunk 1 above)
Comment (zhenggen-xu) [reply cfgmgr/vlanmgr.cpp]: resolved

**Comment by zhenggen-xu** [comment]
@qiluo-msft  Feedback was addressed.

**Pull request merged**